### PR TITLE
feat(dev, node-server): support a hook to access Node.js server instance

### PR DESCRIPTION
**Pull request by OneCalmCloud**
### 🔗 Linked issue
solve https://github.com/unjs/nitro/issues/2040

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [ ]  🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Hi 👋 this PR can be possible to enable WebSocket and Web Server to share the same port without the need for reverse proxy.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/runtime/entries/nitro-dev.ts [hunk 1]
```diff
@@ -47,6 +47,7 @@ const listener = server.listen(listenAddress, () => {
         ? { socketPath: _address }
         : { host: "localhost", port: _address.port },
   });
+  nitroApp.hooks.callHook("server", server);
```
Comment (pi0) [inline src/runtime/entries/nitro-dev.ts:50]: This can be renamed to `dev:server`

File: src/runtime/entries/node-server.ts [hunk 2]
```diff
@@ -42,6 +42,7 @@ const listener = server.listen(path ? { path } : { port, host }, (err) => {
       ? `[${addressInfo.address}]`
       : addressInfo.address
   }:${addressInfo.port}${baseURL}`;
+  nitroApp.hooks.callHook("server", server);
```
Comment (pi0) [inline src/runtime/entries/node-server.ts:45]: This can be renamed to `node:server`

**Review by pi0 — COMMENTED** [review]
Thanks for the PR. Altough main usage of WebSocket integrtion should be solved soon by having built-in websocket layer (which is so frustrating now!) but i think it would be always good idea to be able access listening server.

(An alternative approach to hooks could also be injecting to nitroApp context to access from plugins)

File: src/runtime/entries/nitro-dev.ts (hunk 1 above)
Comment (OneCalmCloud) [reply src/runtime/entries/nitro-dev.ts]: Yes, I renamed to `dev:server`

File: src/runtime/entries/node-server.ts (hunk 2 above)
Comment (OneCalmCloud) [reply src/runtime/entries/node-server.ts]: Yes, I renamed to `node:server`

**Comment by JCtapuk** [comment]
When will this PR be added to the main branch?

**Comment by JCtapuk** [comment]
I propose to expand these events
```ts
const server =
  cert && key
    ? new HttpsServer({ key, cert }, toNodeListener(nitroApp.h3App))
    : new HttpServer(toNodeListener(nitroApp.h3App));
    
// others code
nitroApp.hooks.callHook("node.server.init", server);

server.on('upgrade', () => {
   nitroApp.hooks.callHook("node.server.upgrade", server, req: IncomingMessage);
})

server.on('listening', () => {
   nitroApp.hooks.callHook("node.server.listening", server);
})
```

So much more stable for more control. For example, a handshake for WebSocket and the event that the server has started. As well as initialization which will make it possible to register more events for advanced users

**Comment by OneCalmCloud** [comment]
> When will this PR be added to the main branch?

I'm so sorry for the delay in progressing the PR due to my unfamiliarity with GitHub functionalities; I'm actively learning and seeking guidance to contribute effectively to the project.

**Comment by pi0** [comment]
Sorry for delay on this. I guess now that we have official [WebSocket](https://nitro.unjs.io/guide/websocket) support, it should be less priority but I will defneitly investigate how to make it an extendable addon before introducing 👍🏼

**Pull request closed without merging**